### ISSUE TITLE
Boolean should be on type Field

### ISSUE THREAD

**Issue by DanielFerrariR** [jackfranklin/test-data-bot#209]
Boolean should be on type Field

**Comment by aaronmcadam** [comment]
I can't use a `Date` object either.

**Comment by aaronmcadam** [comment]
Hey @DanielFerrariR, check out this: https://github.com/jackfranklin/test-data-bot#bool

**Comment by jackfranklin** [comment]
Thanks for this bug report. I'm not sure how best to do this because the hope was that with TypeScript you'd get good bug reporting if you pass invalid fields; so I don't want to allow any value to be a field because we'd lose that, but maybe I should explicitly list things like `Date` on here. I'm not sure to be honest.

@DanielFerrariR I think that you could use `perBuild(() => new Date('...'))` to work around it.

**Comment by aaronmcadam** [comment]
@jackfranklin I think we'd have to use generics to enable things like this in a general way.

I had to do the following:

```
const logEvent = logEventBuilder({
    overrides: {
      id: 'abcd',
      level: LogEventLevel.Info,
      message: 'Message',
      meta: {
        jobName: 'Importer',
        hostname: 'Macbook.local',
      },
    },
    map: (logEvent) => {
      logEvent.time = new Date('2020-02-24T14:00:00+0000');

      return logEvent;
    },
  });
```

**Comment by aaronmcadam** [comment]
I managed to get the date working when I don't care about the value, it's only when I want to set it to a specific value when testing outputs:

```ts
export interface LogEvent {
  id: string;
  level: LogEventLevel;
  message: string;
  time: Date;
  meta?: LogEventMeta;
}

export const logEventBuilder = build<LogEvent>('LogEvent', {
  fields: {
    id: sequence(),
    level: LogEventLevel.Error,
    message: fake((f) => f.lorem.paragraphs()),
    time: fake((f) => f.date.recent(1)),
    meta: {
      jobName: fake((f) => f.hacker.noun()),
      hostname: fake((f) => f.internet.domainName()),
    },
  },
});
```

**Comment by DanielFerrariR** [comment]
@jackfranklin

Normally people use generics on lib functions to work around this kind of issue. 

I think build should type any parameter as Field which has the fake function type and also provide a generic type for other users give their types like boolean, date , etc.

**Comment by jackfranklin** [comment]
@DanielFerrariR @aaronmcadam I'd really like to take some time to improve the typings here because it's one big hack and if we extend the `Field` type we'll end up with it just being a list of every type it could ever be.

I wonder about making field something like:

```js
type Field<X> = Sequence<X> | Faker<X> | OneOf<X> | PerBuild<X> | Value<X>
```

And then if you didn't want to use any of the generators you'd use the `value` generator which effectively would spit out whatever you pass it:

```
const user = build({
  fields: { name: value('jack') }
})
```

The `value` generator would have a generic `X` too so TS would be able to infer types (like dates, etc), and then we wouldn't have issues like this one. WDYT?

**Comment by DanielFerrariR** [comment]
Seems like a good idea.

**Comment by jackfranklin** [comment]
@DanielFerrariR yeah that's my thinking. I might branch and spike it and see what happens.

I'm quite keen to avoid a breaking change but I'm not sure if that'll be possible.

**Comment by aaronmcadam** [comment]
This looks like a good idea, but I wonder if could use the `infer` keyword at all to help us.

**Comment by jackfranklin** [comment]
@aaronmcadam please could you give an example of how that might help? I'm not super familiar with how it works and therefore I'm struggling to apply it to test-data-bot.

**Comment by aaronmcadam** [comment]
@jackfranklin I've not used `infer` much myself yet, but it can be used in conditional types to try to avoid needing a more heavy handed approach (https://dev.to/aexol/typescript-tutorial-infer-keyword-2cn). I'm trying to think of a way we can avoid needing an identity function like the `value` one above.

**Comment by mwarger** [comment]
@DanielFerrariR I've gotten around this by using `oneOf(false)` for example. 

But I agree, I think you should be able to just put false if you want.

**Comment by Glinkis** [comment]
This should be fixed in https://github.com/jackfranklin/test-data-bot/pull/599
